text/0005-application-formalization.md

```diff
@@ -0,0 +1,196 @@
+
+Example w/ many version omitted for clarity:
+**app-foo-1.0.1**
+* requires: 
+  * `app-bar: ^1.0.0`
+* modules:
+  * `ui-foo`
+    * requires interfaces: `foo`
+  * `mod-foo`
+    * provides interfaces: `foo`
+    * requires interfaces: `foo-storage`, `bar`
+  * `mod-foo-storage`
+    * provides interfaces: `foo-storage`
+    * requires interfaces: 
+
+**app-bar-1.0.0**
+* requires: 
+  * `app-configuration: ~1.1.1`
+* modules:
+  * `mod-bar`
+    * provides interfaces: `bar`
+    * requires interfaces: `configuration`
+  * `ui-bar`
+    * requires interfaces: `bar`
+
+**app-configuration-1.1.1**
+* requires: 
+* modules:
+  * `mod-configuration`
+    * provides interfaces: `configuration`
+    * requires interfaces: 
+
+In this scenario, in order for `app-foo-1.0.1` to be enabled for a tenant, `app-bar-1.X.X` and `app-configuraturation-1.1.X` would also need to be enabled for the tenant.  This is because `app-foo` depends on a compatible version (specified as `^`) of `app-bar-1.X.X`, and `app-bar` depends on an approximately equivalent version (specified as `~`) of `app-configuration-1.1.1`.  Exact versions and/or ranges can also be specified.  See the [node-semver documentation](https://github.com/npm/node-semver#advanced-range-syntax]) for details.
+
+#### Application Descriptors
+Similar to how modules are explicitly defind through module descriptors, applications are defined through application descriptors.  The application descriptor references module descriptors
+
+**Application Descriptor Properties**
+| Property            | Type         | Description                                                                          | Notes                    |
+| ------------------- | ------------ | ------------------------------------------------------------------------------------ | ------------------------ |
+| id                  | String       | Identifier for the application conforming to pattern: `{name}-{semantic version}`    | e.g. "app-orders-1.0.0"  |
+| name                | String       | Name of the application                                                              | e.g. "app-orders"        |
+| version             | String       | Version of the application                                                           | e.g. "1.0.0"             |
+| description         | String       | Brief description of the application                                                 | e.g. "Application delivering orders functionality" |
+| dependencies        | Dependency[] | List of dependencies on other applications                                           | See "Dependency Properties" table below            |
```

Comment:
Should the optional dependencies also be specified there? For example some application A could interact with another application B. If application B is not installed, the functionality related to the B will be hidden in A.

Reply:
Good point @psmagin!  I'll try to get this worked into the RFC.

Reply:
@psmagin after thinking through this and a bit of discussion with others we can't really think of any use cases for this.  The optional dependencies are still captured in the module descriptors, which aren't going away.  Of course, if there's a good reason to do this we can revisit later.  Thanks again for the suggestion!